File: contracts/activity.mjs
import { validateActuationStream } from "./actuation-stream.mjs";

export const ACTIVITY_VERSION = "actuation.activity/v1";

const PHASES = new Set([
  "queued",
  "running",
  "waiting",
  "completed",
  "failed",
  "interrupted",
  "cancelled",
]);
const OUTCOMES = new Set(["pending", "succeeded", "failed", "degraded", "refused", "cancelled"]);
const SALIENCE = new Set(["ambient", "normal", "important", "critical"]);

function object(value, name) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError(`${name} must be an object`);
  }
  return value;
}

function ref(value, name, { optional = false } = {}) {
  if (value == null && optional) return;
  if (typeof value !== "string" || value.trim() === "") {
    throw new TypeError(`${name} must be a non-empty ref`);
  }
}

function refs(value, name, { optional = false } = {}) {
  if (value == null && optional) return [];
  if (!Array.isArray(value) || value.some((item) => typeof item !== "string" || item.trim() === "")) {
    throw new TypeError(`${name} must be an array of non-empty refs`);
  }
  return value;
}

function text(value, name) {
  if (typeof value !== "string" || value.trim() === "") {
    throw new TypeError(`${name} must be a non-empty string`);
  }
}

function timestamp(value, name, { optional = false } = {}) {
  if (value == null && optional) return;
  if (typeof value !== "string" || value.trim() === "" || Number.isNaN(Date.parse(value))) {
    throw new TypeError(`${name} must be an ISO-compatible timestamp string`);
  }
}

function validateActor(value) {
  const actor = object(value, "Activity.actor");
  ref(actor.agent_ref, "Activity.actor.agent_ref", { optional: true });
  ref(actor.agency_ref, "Activity.actor.agency_ref", { optional: true });
  ref(actor.participant_ref, "Activity.actor.participant_ref", { optional: true });
  ref(actor.locus_ref, "Activity.actor.locus_ref", { optional: true });
  if (actor.agent_ref == null && actor.agency_ref == null && actor.participant_ref == null && actor.locus_ref == null) {
    throw new TypeError("Activity.actor must attribute activity to an Agent, Agency, Participant or locus");
  }
  return actor;
}

function validateTrace(value) {
  const trace = object(value, "Activity.trace");
  ref(trace.stream_ref, "Activity.trace.stream_ref");
  refs(trace.event_refs, "Activity.trace.event_refs", { optional: true });
  refs(trace.native_trace_refs, "Activity.trace.native_trace_refs", { optional: true });
  if (trace.from_sequence != null && (!Number.isSafeInteger(trace.from_sequence) || trace.from_sequence < 1)) {
    throw new TypeError("Activity.trace.from_sequence must be a positive safe integer");
  }
  if (trace.through_sequence != null && (!Number.isSafeInteger(trace.through_sequence) || trace.through_sequence < 1)) {
    throw new TypeError("Activity.trace.through_sequence must be a positive safe integer");
  }
  if (
    trace.from_sequence != null &&
    trace.through_sequence != null &&
    trace.through_sequence < trace.from_sequence
  ) {
    throw new TypeError("Activity.trace.through_sequence must not precede from_sequence");
  }
  return trace;
}

export function validateActivity(input) {
  const activity = object(input, "Activity");
  if (activity.schema !== ACTIVITY_VERSION) {
    throw new TypeError(`Activity.schema must equal ${ACTIVITY_VERSION}`);
  }
  ref(activity.activity_ref, "Activity.activity_ref");
  validateActor(activity.actor);
  ref(activity.agent_session_ref, "Activity.agent_session_ref", { optional: true });
  ref(activity.subject_ref, "Activity.subject_ref");
  ref(activity.native_owner, "Activity.native_owner");
  ref(activity.action_ref, "Activity.action_ref", { optional: true });
  ref(activity.invocation_ref, "Activity.invocation_ref", { optional: true });
  ref(activity.actuation_ref, "Activity.actuation_ref", { optional: true });
  ref(activity.result_ref, "Activity.result_ref", { optional: true });
  ref(activity.return_ref, "Activity.return_ref", { optional: true });
  refs(activity.evidence_refs, "Activity.evidence_refs", { optional: true });
  text(activity.verb, "Activity.verb");
  text(activity.object, "Activity.object");
  text(activity.summary, "Activity.summary");
  if (!PHASES.has(activity.phase)) {
    throw new TypeError("Activity.phase must name a supported semantic phase");
  }
  if (!OUTCOMES.has(activity.outcome)) {
    throw new TypeError("Activity.outcome must name a supported semantic outcome");
  }
  if (!SALIENCE.has(activity.salience)) {
    throw new TypeError("Activity.salience must be ambient, normal, important or critical");
  }
  if (typeof activity.needs_attention !== "boolean") {
    throw new TypeError("Activity.needs_attention must be boolean");
  }
  validateTrace(activity.trace);
  timestamp(activity.started_at, "Activity.started_at");
  timestamp(activity.updated_at, "Activity.updated_at");
  timestamp(activity.completed_at, "Activity.completed_at", { optional: true });
  if (["completed", "failed", "interrupted", "cancelled"].includes(activity.phase) && activity.completed_at == null) {
    throw new TypeError("terminal Activity.phase requires completed_at");
  }
  if (["queued", "running", "waiting"].includes(activity.phase) && activity.completed_at != null) {
    throw new TypeError("non-terminal Activity.phase must not declare completed_at");
  }
  if (activity.metadata != null) object(activity.metadata, "Activity.metadata");
  return activity;
}

function phaseFromStream(stream) {
  switch (stream.lifecycle.state) {
    case "open":
      return "running";
    case "closed":
      return "completed";
    case "interrupted":
      return "interrupted";
    case "cancelled":
      return "cancelled";
    default:
      throw new TypeError(`unsupported ActuationStream lifecycle ${stream.lifecycle.state}`);
  }
}

function outcomeFromPhase(phase) {
  switch (phase) {
    case "completed":
      return "succeeded";
    case "cancelled":
      return "cancelled";
    case "interrupted":
      return "degraded";
    default:
      return "pending";
  }
}

function actorFromEvent(stream, event) {
  const actor = event?.actor ?? {};
  const resolved = {
    agent_ref: actor.agent_ref,
    agency_ref: actor.agency_ref ?? stream.agency_ref,
    participant_ref: actor.participant_ref,
    locus_ref: actor.locus_ref,
  };
  return Object.fromEntries(Object.entries(resolved).filter(([, value]) => value != null));
}

/**
 * Project one semantic Activity over an existing canonical ActuationStream.
 * This does not mutate the stream and does not create Notification/Attention state.
 */
export function activityFromActuationStream(
  streamInput,
  {
    activityRef,
    subjectRef,
    nativeOwner,
    verb,
    object: activityObject,
    summary,
    salience = "normal",
    needsAttention = false,
    actionRef,
    invocationRef,
    resultRef,
    evidenceRefs = [],
    returnRef,
    outcome,
    metadata,
  },
) {
  const stream = validateActuationStream(streamInput);
  const lastEvent = stream.events.at(-1);
  const phase = phaseFromStream(stream);
  const startedAt = stream.lifecycle.started_at ?? stream.events.at(0)?.observed_at;
  const updatedAt = stream.lifecycle.ended_at ?? lastEvent?.observed_at ?? startedAt;
  if (startedAt == null || updatedAt == null) {
    throw new TypeError("Activity projection requires stream lifecycle/event timestamps");
  }

  const activity = {
    schema: ACTIVITY_VERSION,
    activity_ref: activityRef,
    actor: actorFromEvent(stream, lastEvent),
    agent_session_ref: stream.agent_session_ref,
    subject_ref: subjectRef,
    native_owner: nativeOwner,
    action_ref: actionRef,
    invocation_ref: invocationRef,
    actuation_ref: stream.actuation_ref,
    result_ref: resultRef,
    evidence_refs: [...evidenceRefs],
    return_ref: returnRef ?? lastEvent?.return_ref,
    verb,
    object: activityObject,
    summary,
    phase,
    outcome: outcome ?? outcomeFromPhase(phase),
    salience,
    needs_attention: needsAttention,
    trace: {
      stream_ref: stream.stream_ref,
      event_refs: stream.events.map((event) => event.event_ref),
      native_trace_refs: stream.events.flatMap((event) => event.native_trace_ref ? [event.native_trace_ref] : []),
      ...(stream.events.length > 0 ? { from_sequence: 1, through_sequence: stream.events.length } : {}),
    },
    started_at: startedAt,
    updated_at: updatedAt,
    ...(phase === "completed" || phase === "interrupted" || phase === "cancelled"
      ? { completed_at: updatedAt }
      : {}),
    ...(metadata == null ? {} : { metadata: structuredClone(metadata) }),
  };
  return validateActivity(activity);
}

/**
 * Project a single meaningful portable stream event as semantic Activity. It is
 * valid for the event to have no Action/Invocation identity: observation is not
 * counterfeited into an Action merely to make it visible.
 */
export function activityFromStreamEvent(
  streamInput,
  eventRef,
  {
    activityRef,
    subjectRef,
    nativeOwner,
    verb,
    object: activityObject,
    summary,
    phase = "completed",
    outcome = "succeeded",
    salience = "normal",
    needsAttention = false,
    actionRef,
    invocationRef,
    resultRef,
    evidenceRefs = [],
    returnRef,
  },
) {
  const stream = validateActuationStream(streamInput);
  const event = stream.events.find((candidate) => candidate.event_ref === eventRef);
  if (!event) throw new TypeError(`ActuationStream contains no event ${eventRef}`);
  const observedAt = event.observed_at ?? stream.lifecycle.started_at;
  if (observedAt == null) throw new TypeError("Activity event projection requires a timestamp");

  return validateActivity({
    schema: ACTIVITY_VERSION,
    activity_ref: activityRef,
    actor: actorFromEvent(stream, event),
    agent_session_ref: stream.agent_session_ref,
    subject_ref: subjectRef,
    native_owner: nativeOwner,
    action_ref: actionRef,
    invocation_ref: invocationRef,
    actuation_ref: stream.actuation_ref,
    result_ref: resultRef,
    evidence_refs: [...evidenceRefs, ...(event.evidence_refs ?? [])],
    return_ref: returnRef ?? event.return_ref,
    verb,
    object: activityObject,
    summary,
    phase,
    outcome,
    salience,
    needs_attention: needsAttention,
    trace: {
      stream_ref: stream.stream_ref,
      event_refs: [event.event_ref],
      native_trace_refs: event.native_trace_ref ? [event.native_trace_ref] : [],
      from_sequence: event.sequence,
      through_sequence: event.sequence,
    },
    started_at: observedAt,
    updated_at: observedAt,
    ...(["completed", "failed", "interrupted", "cancelled"].includes(phase)
      ? { completed_at: observedAt }
      : {}),
  });
}

export function activityNeedsAttention(activityInput) {
  return validateActivity(activityInput).needs_attention;
}

File: crates/actuation-stream/src/activity.rs
use crate::{domain, Count, JsonObject, Timestamp};
use crate::{ActuationStream, Attribution, AttributionFields, StreamEvent, StreamState};
use actuation_core::*;
use serde::{Deserialize, Serialize};
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum ActivitySchema {
    #[serde(rename = "actuation.activity/v1")]
    V1,
}
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum ActivityPhase {
    #[serde(rename = "queued")]
    Queued,
    #[serde(rename = "running")]
    Running,
    #[serde(rename = "waiting")]
    Waiting,
    #[serde(rename = "completed")]
    Completed,
    #[serde(rename = "failed")]
    Failed,
    #[serde(rename = "interrupted")]
    Interrupted,
    #[serde(rename = "cancelled")]
    Cancelled,
}
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum ActivityOutcome {
    #[serde(rename = "pending")]
    Pending,
    #[serde(rename = "succeeded")]
    Succeeded,
    #[serde(rename = "failed")]
    Failed,
    #[serde(rename = "degraded")]
    Degraded,
    #[serde(rename = "refused")]
    Refused,
    #[serde(rename = "cancelled")]
    Cancelled,
}
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum Salience {
    #[serde(rename = "ambient")]
    Ambient,
    #[serde(rename = "normal")]
    Normal,
    #[serde(rename = "important")]
    Important,
    #[serde(rename = "critical")]
    Critical,
}
impl ActivityPhase {
    pub fn is_terminal(self) -> bool {
        matches!(
            self,
            Self::Completed | Self::Failed | Self::Interrupted | Self::Cancelled
        )
    }
}
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ActivityTraceFields {
    pub stream_ref: StreamRef,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub event_refs: Slot<Vec<EventRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub native_trace_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub from_sequence: Slot<Count>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub through_sequence: Slot<Count>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
domain!(
    ActivityTrace,
    ActivityTraceFields,
    [
        stream_ref,
        event_refs,
        native_trace_refs,
        from_sequence,
        through_sequence
    ],
    |v: &Self| {
        if v.from_sequence.value() == Some(&Count::ZERO)
            || v.through_sequence.value() == Some(&Count::ZERO)
        {
            return Err(Error::new("trace sequence must be positive"));
        }
        if let (Some(from), Some(through)) = (v.from_sequence.value(), v.through_sequence.value()) {
            if through < from {
                return Err(Error::new("trace end must not precede its beginning"));
            }
        }
        Ok(())
    }
);

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ActivityFields {
    pub schema: ActivitySchema,
    pub activity_ref: ActivityRef,
    pub actor: Attribution,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub agent_session_ref: Slot<AgentSessionRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub plan_ref: Slot<PlanRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub journey_ref: Slot<JourneyRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub run_ref: Slot<RunRef>,
    pub subject_ref: ExternalRef,
    pub native_owner: ExternalRef,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub action_ref: Slot<ActionRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub invocation_ref: Slot<InvocationRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub actuation_ref: Slot<ActuationRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub result_ref: Slot<ExternalRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub return_ref: Slot<ReturnRef>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub evidence_refs: Slot<Vec<ExternalRef>>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub usage_refs: Slot<Vec<ExternalRef>>,
    pub verb: ExternalRef,
    pub object: ExternalRef,
    pub summary: ExternalRef,
    pub phase: ActivityPhase,
    pub outcome: ActivityOutcome,
    pub salience: Salience,
    pub needs_attention: bool,
    pub trace: ActivityTrace,
    pub started_at: Timestamp,
    pub updated_at: Timestamp,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub completed_at: Slot<Timestamp>,
    #[serde(default, skip_serializing_if = "Slot::is_absent")]
    pub metadata: Slot<JsonObject>,
    #[serde(flatten)]
    pub extensions: Extensions,
}
domain!(
    Activity,
    ActivityFields,
    [
        schema,
        activity_ref,
        actor,
        agent_session_ref,
        plan_ref,
        journey_ref,
        run_ref,
        subject_ref,
        native_owner,
        action_ref,
        invocation_ref,
        actuation_ref,
        result_ref,
        return_ref,
        evidence_refs,
        usage_refs,
        verb,
        object,
        summary,
        phase,
        outcome,
        salience,
        needs_attention,
        trace,
        started_at,
        updated_at,
        completed_at,
        metadata
    ],
    |v: &Self| {
        if v.phase.is_terminal() != v.completed_at.value().is_some() {
            Err(Error::new(
                "terminal Activity requires completed_at; non-terminal Activity forbids it",
            ))
        } else {
            Ok(())
        }
    }
);

/// Supplied semantic descriptions of an observed act; none of these fields
/// discovers caller identity, creates an Action, or establishes authority.
#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityDescription {
    pub activity_ref: ActivityRef,
    pub subject_ref: ExternalRef,
    pub native_owner: ExternalRef,
    pub verb: ExternalRef,
    pub object: ExternalRef,
    pub summary: ExternalRef,
    #[serde(default = "normal_salience")]
    pub salience: Salience,
    #[serde(default)]
    pub needs_attention: bool,
    #[serde(default)]
    pub action_ref: Slot<ActionRef>,
    #[serde(default)]
    pub invocation_ref: Slot<InvocationRef>,
    #[serde(default)]
    pub plan_ref: Option<PlanRef>,
    #[serde(default)]
    pub journey_ref: Option<JourneyRef>,
    #[serde(default)]
    pub run_ref: Option<RunRef>,
    #[serde(default)]
    pub result_ref: Slot<ExternalRef>,
    #[serde(default)]
    pub return_ref: Slot<ReturnRef>,
    #[serde(default)]
    pub evidence_refs: Vec<ExternalRef>,
    #[serde(default)]
    pub outcome: Slot<ActivityOutcome>,
    #[serde(default)]
    pub metadata: Slot<JsonObject>,
}
fn normal_salience() -> Salience {
    Salience::Normal
}
fn optional<T>(value: Option<T>) -> Slot<T> {
    value.map(Slot::Value).unwrap_or(Slot::Absent)
}
fn actor_for(stream: &ActuationStream, event: Option<&StreamEvent>) -> Result<Attribution> {
    let actor = event
        .and_then(|e| e.fields().actor.value())
        .map(Attribution::fields);
    Attribution::new(AttributionFields {
        agent_ref: optional(actor.and_then(|a| a.agent_ref.value()).cloned()),
        agency_ref: Slot::Value(
            actor
                .and_then(|a| a.agency_ref.value())
                .cloned()
                .unwrap_or_else(|| stream.fields().agency_ref.clone()),
        ),
        participant_ref: optional(actor.and_then(|a| a.participant_ref.value()).cloned()),
        locus_ref: optional(actor.and_then(|a| a.locus_ref.value()).cloned()),
        extensions: Extensions::new(),
    })
}
impl Activity {
    pub fn needs_attention(&self) -> bool {
        self.fields().needs_attention
    }
    pub fn from_stream(stream: &ActuationStream, d: ActivityDescription) -> Result<Self> {
        let s = stream.fields();
        let last = s.events.last();
        let phase = match s.lifecycle.fields().state {
            StreamState::Open => ActivityPhase::Running,
            StreamState::Closed => ActivityPhase::Completed,
            StreamState::Interrupted => ActivityPhase::Interrupted,
            StreamState::Cancelled => ActivityPhase::Cancelled,
        };
        let outcome = d.outcome.value().copied().unwrap_or(match phase {
            ActivityPhase::Completed => ActivityOutcome::Succeeded,
            ActivityPhase::Cancelled => ActivityOutcome::Cancelled,
            ActivityPhase::Interrupted => ActivityOutcome::Degraded,
            _ => ActivityOutcome::Pending,
        });
        let started = s
            .lifecycle
            .fields()
            .started_at
            .value()
            .or_else(|| {
                s.events
                    .first()
                    .and_then(|e| e.fields().observed_at.value())
            })
            .ok_or_else(|| {
                Error::new("Activity projection requires stream lifecycle/event timestamps")
            })?;
        let updated = s
            .lifecycle
            .fields()
            .ended_at
            .value()
            .or_else(|| last.and_then(|e| e.fields().observed_at.value()))
            .unwrap_or(started);
        let mut usages = Vec::new();
        for e in &s.events {
            if let Some(u) = e.fields().model_usage.value() {
                if !usages.contains(&u.fields().usage_ref) {
                    usages.push(u.fields().usage_ref.clone());
                }
            }
        }
        let trace = ActivityTrace::new(ActivityTraceFields {
            stream_ref: s.stream_ref.clone(),
            event_refs: Slot::Value(
                s.events
                    .iter()
                    .map(|e| e.fields().event_ref.clone())
                    .collect(),
            ),
            native_trace_refs: Slot::Value(
                s.events
                    .iter()
                    .filter_map(|e| e.fields().native_trace_ref.value().cloned())
                    .collect(),
            ),
            from_sequence: if s.events.is_empty() {
                Slot::Absent
            } else {
                Slot::Value(Count::ONE)
            },
            through_sequence: if s.events.is_empty() {
                Slot::Absent
            } else {
                Slot::Value(Count::new(s.events.len() as u64)?)
            },
            extensions: Extensions::new(),
        })?;
        let return_ref = d
            .return_ref
            .value()
            .cloned()
            .map(Slot::Value)
            .unwrap_or_else(|| {
                last.map(|e| e.fields().return_ref.clone())
                    .unwrap_or(Slot::Absent)
            });
        let metadata = optional(d.metadata.value().cloned());
        Self::new(ActivityFields {
            schema: ActivitySchema::V1,
            activity_ref: d.activity_ref,
            actor: actor_for(stream, last)?,
            agent_session_ref: Slot::Value(s.agent_session_ref.clone()),
            plan_ref: optional(d.plan_ref),
            journey_ref: optional(d.journey_ref),
            run_ref: optional(d.run_ref),
            subject_ref: d.subject_ref,
            native_owner: d.native_owner,
            action_ref: d.action_ref,
            invocation_ref: d.invocation_ref,
            actuation_ref: Slot::Value(s.actuation_ref.clone()),
            result_ref: d.result_ref,
            return_ref,
            evidence_refs: Slot::Value(d.evidence_refs),
            usage_refs: if usages.is_empty() {
                Slot::Absent
            } else {
                Slot::Value(usages)
            },
            verb: d.verb,
            object: d.object,
            summary: d.summary,
            phase,
            outcome,
            salience: d.salience,
            needs_attention: d.needs_attention,
            trace,
            started_at: started.clone(),
            updated_at: updated.clone(),
            completed_at: if phase.is_terminal() {
                Slot::Value(updated.clone())
            } else {
                Slot::Absent
            },
            metadata,
            extensions: Extensions::new(),
        })
    }
    pub fn from_event(
        stream: &ActuationStream,
        event_ref: &EventRef,
        d: ActivityDescription,
        phase: ActivityPhase,
    ) -> Result<Self> {
        let s = stream.fields();
        let event = s
            .events
            .iter()
            .find(|e| &e.fields().event_ref == event_ref)
            .ok_or_else(|| Error::new("ActuationStream contains no such event"))?;
        let e = event.fields();
        let observed = e
            .observed_at
            .value()
            .or(s.lifecycle.fields().started_at.value())
            .ok_or_else(|| Error::new("Activity event projection requires a timestamp"))?;
        let outcome = match d.outcome {
            Slot::Value(v) => v,
            Slot::Absent => ActivityOutcome::Succeeded,
            Slot::Null => return Err(Error::new("Activity outcome cannot be null")),
        };
        let mut evidence = d.evidence_refs;
        evidence.extend(e.evidence_refs.value().cloned().unwrap_or_default());
        let return_ref = d
            .return_ref
            .value()
            .cloned()
            .map(Slot::Value)
            .unwrap_or_else(|| e.return_ref.clone());
        Self::new(ActivityFields {
            schema: ActivitySchema::V1,
            activity_ref: d.activity_ref,
            actor: actor_for(stream, Some(event))?,
            agent_session_ref: Slot::Value(s.agent_session_ref.clone()),
            plan_ref: optional(d.plan_ref),
            journey_ref: optional(d.journey_ref),
            run_ref: optional(d.run_ref),
            subject_ref: d.subject_ref,
            native_owner: d.native_owner,
            action_ref: d.action_ref,
            invocation_ref: d.invocation_ref,
            actuation_ref: Slot::Value(s.actuation_ref.clone()),
            result_ref: d.result_ref,
            return_ref,
            evidence_refs: Slot::Value(evidence),
            usage_refs: optional(
                e.model_usage
                    .value()
                    .map(|u| vec![u.fields().usage_ref.clone()]),
            ),
            verb: d.verb,
            object: d.object,
            summary: d.summary,
            phase,
            outcome,
            salience: d.salience,
            needs_attention: d.needs_attention,
            trace: ActivityTrace::new(ActivityTraceFields {
                stream_ref: s.stream_ref.clone(),
                event_refs: Slot::Value(vec![e.event_ref.clone()]),
                native_trace_refs: Slot::Value(
                    e.native_trace_ref.value().cloned().into_iter().collect(),
                ),
                from_sequence: Slot::Value(e.sequence),
                through_sequence: Slot::Value(e.sequence),
                extensions: Extensions::new(),
            })?,
            started_at: observed.clone(),
            updated_at: observed.clone(),
            completed_at: if phase.is_terminal() {
                Slot::Value(observed.clone())
            } else {
                Slot::Absent
            },
            metadata: Slot::Absent,
            extensions: Extensions::new(),
        })
    }
}
